client-java/src/main/java/uk/gov/gchq/palisade/client/job/ClientJob.java

```diff
@@ -0,0 +1,362 @@
+     *
+     * @param func The builder function
+     * @return a newly created ClientJob instance
+     */
+    @SuppressWarnings("java:S3242") // I REALLY want to use UnaryOperator here SonarQube!!!
+    public static ClientJob createJob(final UnaryOperator<JobSetup.Builder> func) {
+        checkArgument(func, "Must supply a function to create a job");
+        return createJob(func.apply(JobSetup.builder()).build());
+    }
+
+    /**
+     * Helper method to create a {@code ClientJob}
+     *
+     * @param setup The job setup
+     * @return a newly created ClientJob
+     */
+    public static ClientJob createJob(final JobSetup setup) {
+        checkArgument(setup, "Must supply a the setup for the job");
+        return new ClientJob(setup);
+    }
+
+    /**
+     * Handles the event that signifies that a download has completed normally. The
+     * internal state is updated. If the resource client for this job has completed,
+     * then the number of downloads outstanding is decremented.
+     *
+     * @param event The event to be handled
+     */
+    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
+    @SuppressWarnings("java:S3242")
+    public void onDownloadCompleted(final DownloadCompletedEvent event) {
+        state.downloadCompleted(event.getId(), event.getTime(), event.getProperties());
+        if (this.remainingDownloadLatch != null) {
+            remainingDownloadLatch.countDown();
+            LOGGER.debug("Job is shutting down. Latch decremented");
+        }
+    }
+
+    /**
+     * Handles the event that signifies that a download has completed abnormally.
+     * The internal state is updated. If the resource client for this job has
+     * completed, then the number of downloads outstanding is decremented.
+     *
+     * @param event The event to be handled
+     */
+    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
+    @SuppressWarnings("java:S3242")
+    public void onDownloadFailed(final DownloadFailedEvent event) {
+        state.downloadFailed(event.getId(), event.getTime(), event.getStatusCode(), event.getCause());
+        if (this.remainingDownloadLatch != null) {
+            remainingDownloadLatch.countDown();
+            LOGGER.debug("Job is shutting down. Latch decremented");
+        }
+    }
+
+    /**
+     * Handles the event that signifies that a download has started. The internal
+     * state is updated.
+     *
+     * @param event The event to be handled
+     */
+    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
+    @SuppressWarnings("java:S3242")
+    public void onDownloadStarted(final DownloadStartedEvent event) {
+        state.downloadStarted(event.getId(), event.getTime());
+    }
+
+    /**
+     * Handles the event that signifies that a download has been scheduled. The
+     * internal state is updated.
+     *
+     * @param event The event to be handled
+     */
+    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
+    @SuppressWarnings("java:S3242")
```

Comment:
Cannot remove that warning. The inheritence hierarchy is as follows:
```
IDownloadEvent
| -- IDownScheduledEvent
| -- | -- DownScheduledEvent
```
I cannot use `IDownloadEvent` here, otherwise this method would catch all download events. In this situation SonarQube is wrong. I would say this suppression should stay.

Reply:
Removed